### ISSUE TITLE
Clase 5 Pasaje por Referencia

### ISSUE THREAD

**Issue by gviardinformatica** [DamRCorba/Informatica2#6]

**Comment by BrutenCristian** [comment]
Buenas profe, sobre este tema en adelante subieron algún ejercicio para que hagamos? porque no se si no se subió nada o a mi no me llego nada.
Muchas gracias

**Comment by DamRCorba** [comment]
Se los había planteado en la sesion. Por ejemplo hacer la función inicializar sensores para el ejemplo de telemetría. 
te dejo aquí el vídeo ya resuelto [video](https://youtu.be/Gwi7xPKqozs)

**Comment by BrutenCristian** [comment]
Buenas profe, en el ejercicio 13 cuando imprimo me aparece basura en vez de un espacio en blanco, que puede ser?

**Comment by Ana-Francioni** [comment]
Hola profe, si quiero pasar una matriz a una funcion usando pasaje por referencia, como puedo hacer? Usando vectores los puedo pasar sin problema pero pasando una matriz me tira error al compilar. La matriz que quiero pasar es esta:

```c
	char operaciones[4][15];
	strcpy (operaciones[0], SUMA);
	strcpy (operaciones[1], RESTA);
	strcpy (operaciones[2], MULTIPLICACION);
	strcpy (operaciones[3], DIVISION); 
```
como son varias cadenas de caracteres, la alternativa que se me ocurrio fue hacer esto:

```c
typedef struct {
char operacion [15];
} oper_t;

oper_t operaciones[4];
strcpy (operaciones[0].operacion, SUMA);
strcpy (operaciones[1].operacion, RESTA);
strcpy (operaciones[2].operacion, MULTIPLICACION);
strcpy (operaciones[3].operacion, DIVISION);
```
entonces al ser operaciones un vector lo puedo pasar sin problemas a la funcion y el programa anda, pero se puede hacer de otra forma?

**Comment by DamRCorba** [comment]
> Buenas profe, en el ejercicio 13 cuando imprimo me aparece basura en vez de un espacio en blanco, que puede ser?

Pueden ser muchas cosas. Lo que primero se me ocurre es que rompiste la cadena y eso se hace pisando el carácter nulo del final por otro carácter. Para agregar el espacio, tenes varias formas. La mas simple es con strcat, como hice en el [video](https://youtu.be/U-j4qYOOecg). La otra opción es reemplazar el caracter nulo del final de la cadena por el caracter ' ' pasar a la posición siguiente y cargar el 0 para darle el nuevo fin de cadena.

**Comment by DamRCorba** [comment]
> Hola profe, si quiero pasar una matriz a una funcion usando pasaje por referencia, como puedo hacer? Usando vectores los puedo pasar sin problema pero pasando una matriz me tira error al compilar. La matriz que quiero pasar es esta:
> 
> ```c
> 	char operaciones[4][15];
> 	strcpy (operaciones[0], SUMA);
> 	strcpy (operaciones[1], RESTA);
> 	strcpy (operaciones[2], MULTIPLICACION);
> 	strcpy (operaciones[3], DIVISION); 
> ```
> 
> como son varias cadenas de caracteres, la alternativa que se me ocurrio fue hacer esto:
> 
> ```c
> typedef struct {
> char operacion [15];
> } oper_t;
> 
> oper_t operaciones[4];
> strcpy (operaciones[0].operacion, SUMA);
> strcpy (operaciones[1].operacion, RESTA);
> strcpy (operaciones[2].operacion, MULTIPLICACION);
> strcpy (operaciones[3].operacion, DIVISION);
> ```
> 
> entonces al ser operaciones un vector lo puedo pasar sin problemas a la funcion y el programa anda, pero se puede hacer de otra forma?

Hola Ana, tengo una duda, SUMA es una macro que contiene la cadena "SUMA"? si no es asi, te estan faltando las comillas para indicar que SUMA es una cadena de caracteres.

A la matriz la pasas como si fuese un vector, vas a tener todas las filas seguidas, en tu caso 4 filas de 15 posiciones. En la función que recibís el parámetro podes usar un puntero el cual corras en 15 por cada fila que acceder.

En datos tu matriz se almacenaria asi:
```c
{'S','U','M','A',0,X,X,X,X,X,X,X,X,X,X,'R','E','S','T','A',0,X,X,X,X,X,X,X,X,X,'M','U','L','T','I','P','L','I','C','A','C','I','O','N',0,'D','I','V','I','C','I','O','N',0,X,X,X,X,X,X}
```

**Comment by Ana-Francioni** [comment]
Hola profe, 

En realidad suma y las demás operaciones son constantes que definí antes, por eso no puse las comillas. 
Yo lo que hice fue lo siguiente:

```c

	char operaciones[4][15];
	strcpy (operaciones[0], SUMA);
	strcpy (operaciones[1], RESTA);
	strcpy (operaciones[2], MULTIPLICACION);
	strcpy (operaciones[3], DIVISION);
```
Ahí definí mi matriz, y el prototipo de función que la llama así:

```c
calculadora_t confirmacion (char operaciones[], calculadora_t calculadora)
```
El error que me da al compilar me dice: 

[Warning] passing argument 1 of 'confirmacion' from incompatible pointer type.

Puede ser que el error esté en como declaré el argumento de la función?

**Comment by DamRCorba** [comment]
> Hola profe,
> 
> En realidad suma y las demás operaciones son constantes que definí antes, por eso no puse las comillas.
> Yo lo que hice fue lo siguiente:
> 
> ```c
> 	char operaciones[4][15];
> 	strcpy (operaciones[0], SUMA);
> 	strcpy (operaciones[1], RESTA);
> 	strcpy (operaciones[2], MULTIPLICACION);
> 	strcpy (operaciones[3], DIVISION);
> ```
> 
> Ahí definí mi matriz, y el prototipo de función que la llama así:
> 
> ```c
> calculadora_t confirmacion (char operaciones[], calculadora_t calculadora)
> ```
> 
> El error que me da al compilar me dice:
> 
> [Warning] passing argument 1 of 'confirmacion' from incompatible pointer type.
> 
> Puede ser que el error esté en como declaré el argumento de la función?

Claro, vos tenes que pasarle solo el puntero.
```c
calculadora_t confirmacion (char *operaciones, calculadora_t calculadora)
```

**Comment by Ana-Francioni** [comment]
Ah buenísimo ahora si. Y si por ejemplo en la función quiero imprimir una de las 4 cadenas de la matriz como hago? Porque si lo maneja como un vector a lo largo digamos, si quisiera imprimir la primer cadena y pongo operaciones[0] me imprime solamente el primer carácter y no toda la cadena.

**Comment by DamRCorba** [comment]
Claro, como estas pasando la posicion de memoria del primer elemento de la matriz y esta se guarda como un vector largo, lo que deberias hacer es imprimir con %s en el *p que apunte a la primer posicion. Para la segunda posicion seria *(p+largo del vector) y asi con cada una de las posiciones. 
```c
* p                                     *(p+logitud)                           *(p+(logitud*2))   
{'S','U','M','A',0,X,X,X,X,X,X,X,X,X,X,'R','E','S','T','A',0,X,X,X,X,X,X,X,X,X,'M','U','L','T','I','P','L','I','C','A','C','I','O','N',0,'D','I','V','I','C','I','O','N',0,X,X,X,X,X,X}
```
En *p tenes SUMA
EN *(p+logitud) RESTA
En *(p+(longitud*2)) Multipliacacion
En *(p+(longitud*3)) Division.

**Comment by BrutenCristian** [comment]
Buenas profe, cuando compilo el ejercicio 16 la el codeblocks me tira este error que no me habia aparecido nunca: fatal error: LEY_DE_OHM_H_INCLUDED: No such file or directory|, y la declare como lo venia haciendo siempre

**Comment by DamRCorba** [comment]
Hola Cristian, escribiste y borraste dos preguntas hoy?

    El martes, 16 de junio de 2020 18:00:31 ART, BrutenCristian <notifications@github.com> escribió:  
 
 


Buenas profe, cuando compilo el ejercicio 16 la el codeblocks me tira este error que no me habia aparecido nunca: fatal error: LEY_DE_OHM_H_INCLUDED: No such file or directory|, y la declare como lo venia haciendo siempre

**Comment by BrutenCristian** [comment]
Buenas profe, estaba haciendo un ejercicio y me tiraba algunos errores que no me habían pasado nunca, pero después encontré donde me había equivocado, eran todos errores de tipeo, por eso borre las preguntas
